component/prometheus/operator/podmonitors/crdmanager.go

```diff
@@ -0,0 +1,280 @@
+package podmonitors
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/agent/component"
+	"github.com/grafana/agent/component/prometheus"
+	"github.com/grafana/agent/component/prometheus/operator/podmonitors/config_gen"
+	"github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/discovery"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+	"github.com/prometheus/prometheus/scrape"
+	toolscache "k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	v1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// Generous timeout period for configuring all informers
+const informerSyncTimeout = 10 * time.Second
+
+// CRDManager is all of the fields required to run the component.
+// on update, this entire thing will be recreated and restarted
+type CRDManager struct {
+	Mut              sync.Mutex
+	DiscoveryConfigs map[string]discovery.Configs
+	ScrapeConfigs    map[string]*config.ScrapeConfig
+	DebugInfo        map[string]*DiscoveredPodMonitor
+	DiscoveryManager *discovery.Manager
+	ScrapeManager    *scrape.Manager
+
+	Opts      component.Options
+	Logger    log.Logger
+	Args      *Arguments
+	ConfigGen config_gen.ConfigGenerator
```

Comment:
Why are all of these exported? That feels like a code smell to me, especially when some of them are suppose to be protected by a mutex.